### PR TITLE
feat: heatmap storage and API

### PR DESCRIPTION
**Pull request by pauldambra**
Splitting parts out of #21487 which was a good spike PR but needs breaking up now

This adds 

* the CH table 
* the HogQL wiring
* an API reading heatmaps from the table using HogQL

### CODE REVIEW

File: posthog/heatmaps/sql.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+# This table is responsible for reading from heatmaps on a cluster setting
+DISTRIBUTED_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=HEATMAPS_DATA_TABLE(),
+        # I don't think there's a great natural sharding key here
+        # we'll be querying for team data by url and date,
+        # so I _think_ this offers a reasonable spread of write load
+        # without needing to query too many shards
```
Comment (pauldambra) [inline posthog/heatmaps/sql.py:159]:
```suggestion
```

File: posthog/heatmaps/sql.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+# This table is responsible for writing to sharded_heatmaps based on a sharding key.
+WRITABLE_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="writable_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=HEATMAPS_DATA_TABLE(),
+        # I don't think there's a great natural sharding key here
+        # we'll be querying for team data by url and date,
+        # so I _think_ this offers a reasonable spread of write load
+        # without needing to query too many shards
```
Comment (pauldambra) [inline posthog/heatmaps/sql.py:145]:
```suggestion
```

File: posthog/heatmaps/heatmaps_api.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    count = serializers.IntegerField(required=True)
+    pointer_y = serializers.IntegerField(required=True)
+    pointer_relative_x = serializers.FloatField(required=True)
+    pointer_target_fixed = serializers.BooleanField(required=True)
+
+
+def default_start_date():
+    return (datetime.now() - timedelta(days=7)).date()
+
+
```
Comment (benjackwhite) [inline posthog/heatmaps/heatmaps_api.py:64]:
```suggestion
```

File: posthog/heatmaps/heatmaps_api.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    viewport_width_max = serializers.IntegerField(required=False)
+    type = serializers.CharField(required=False, default="click")
+    date_from = serializers.CharField(required=False, default="-7d")
+    date_to = serializers.DateField(required=False)
+    url_exact = serializers.CharField(required=False)
+    url_pattern = serializers.CharField(required=False)
+    aggregation = serializers.CharField(required=False, default="total_count")
+
+    def validate_aggregation(self, value: str) -> str:
+        if value not in ["total_count", "unique_visitors"]:
```
Comment (benjackwhite) [inline posthog/heatmaps/heatmaps_api.py:76]: This could be a ChoiceField

File: posthog/heatmaps/heatmaps_api.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+        return values
+
+
+class HeatmapsResponseSerializer(serializers.Serializer):
+    results = HeatmapResponseItemSerializer(many=True)
+
+
+class HeatmapsScrollDepthResponseSerializer(serializers.Serializer):
+    results = serializers.ListField(child=serializers.DictField())
```
Comment (benjackwhite) [inline posthog/heatmaps/heatmaps_api.py:112]: Why no serializer?

File: posthog/heatmaps/heatmaps_api.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        if is_scrolldepth_query:
+            aggregation_value = "count(*)" if aggregation == "total_count" else "count(distinct distinct_id)"
+        aggregation_count = parse_expr(aggregation_value)
+
+        exprs = self._predicate_expressions(placeholders)
+
+        stmt = parse_select(raw_query, {"aggregation_count": aggregation_count, "predicates": ast.And(exprs=exprs)})
+
+        context = HogQLContext(team_id=self.team.pk, limit_top_select=False)
+        doohickies = execute_hogql_query(
```
Comment (benjackwhite) [inline posthog/heatmaps/heatmaps_api.py:147]: Fun police - we need a less silly var name here

File: posthog/hogql/database/schema/heatmaps.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        "x": IntegerDatabaseField(name="x"),
+        "y": IntegerDatabaseField(name="y"),
+        "scale_factor": IntegerDatabaseField(name="scale_factor"),
+        "viewport_width": IntegerDatabaseField(name="viewport_width"),
+        "viewport_height": IntegerDatabaseField(name="viewport_height"),
+        "pointer_target_fixed": BooleanDatabaseField(name="pointer_target_fixed"),
+        "current_url": StringDatabaseField(name="current_url"),
+        "timestamp": DateTimeDatabaseField(name="timestamp"),
+        "type": StringDatabaseField(name="type"),
+        # join to get analytics for heatmap areas?
```
Comment (benjackwhite) [inline posthog/hogql/database/schema/heatmaps.py:27]: Feels like this should be removed or implemented properly

**Review by benjackwhite — CHANGES_REQUESTED** [review]
A few things - @pauldambra let me know if you want me to implement the changes suggested

**Comment by pauldambra** [comment]
🤣 we both did it...

good test of how consistent a change would be if we both did it 🤣

**Review by benjackwhite — APPROVED** [review]

**Comment by pauldambra** [comment]
paging @fuziontech 😊 

plz can we do these things to the clickhouse

File: posthog/heatmaps/sql.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    y Int16,
+    -- stored so that in future we can support other resolutions
+    scale_factor Int16,
+    viewport_width Int16,
+    viewport_height Int16,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed Bool,
+    current_url VARCHAR,
+    type LowCardinality(String),
+    _timestamp DateTime
```
Comment (fuziontech) [inline posthog/heatmaps/sql.py:68]: something else that would be _lovely_ to add here are the following [virtual columns](https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka#virtual-columns):
`_offset`
`_partition`

We've really wished to have these on tables in the past - would be super cool to start adding them here.

File: posthog/heatmaps/sql.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    PARTITION BY toYYYYMM(timestamp)
+    -- almost always this is being queried by
+    --   * type,
+    --   * team_id,
+    --   * date range,
+    --   * URL (maybe matching wild cards),
+    --   * width
+    -- we'll almost never query this by session id
+    -- so from least to most cardinality that's
+    ORDER BY (type, team_id,  toDate(timestamp), current_url, viewport_width)
```
Comment (fuziontech) [inline posthog/heatmaps/sql.py:86]: love it, this is perfect. Well done 🙏

File: posthog/heatmaps/sql.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    y,
+    -- stored so that in future we can support other resolutions
+    scale_factor,
+    viewport_width,
+    viewport_height,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed,
+    current_url,
+    type,
+    _timestamp
```
Comment (fuziontech) [inline posthog/heatmaps/sql.py:125]: same thing here:
something else that would be lovely to add here are the following [virtual columns](https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka#virtual-columns):
_offset
_partition

We've really wished to have these on tables in the past - would be super cool to start adding them here.

File: posthog/heatmaps/sql.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+DROP_HEATMAPS_TABLE_SQL = lambda: (
+    f"DROP TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
+)
+
+TRUNCATE_HEATMAPS_TABLE_SQL = lambda: (
+    f"TRUNCATE TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
+)
+
+INSERT_SINGLE_HEATMAP_EVENT = """
```
Comment (fuziontech) [inline posthog/heatmaps/sql.py:164]: This is a scary bit of text - as long as it's only ever used for testing we are good!

**Review by fuziontech — APPROVED** [review]
Thanks for the reminder here @pauldambra 🙏 I needed the bump from last week so I appreciate that. This looks great - only one request to add a few kafka virtual columns, but that's totally optional.

image

**Comment by pauldambra** [comment]
@benjackwhite I have confirmed the migration and ingestion with these latest changes locally (well, I was on partition 0 but the offset was populating)

am in and out all morning but feel free to merge without waiting for me - or I'll do it this afternoon

**Comment by pauldambra** [comment]
* primed kafka setup on master
* then switched to this branch and migrated
* and could see the heatmap

Screenshot 2024-04-23 at 11 14 12

**Pull request merged**